
<img src="https://i.ibb.co/1T5Wbn6/Banner-para-Linkedin-capa-de-perfil-para-programador.png"  alt="KCode Banner"  />

<a href="https://www.linkedin.com/in/kauebr/"><img src="https://cdn-icons-png.flaticon.com/512/174/174857.png" width="30" ></a> <a href="mailto: kauemandarino@gmail.com"><img src="https://cdn-icons-png.flaticon.com/512/5968/5968534.png" width="30" ></a><a href="
https://api.whatsapp.com/send?phone=5541991835800&text=Oi"> <img src="https://imagepng.org/wp-content/uploads/2017/08/whatsapp-icone-1.png" width="30" ></a><a href="https://github.com/kauebr/"><img src="https://github.githubassets.com/images/modules/logos_page/Octocat.png" width="30" ></a>


# Nascidos Vivos 2020
## Reduzindo dataset
Dados minerados atráves do DATASUS(SINASC, Sistema de Informações Sobre Nascidos Vivos), através da biblioteca [PYSUS](https://github.com/fccoelho/PySUS), utilizando o Google Colab, seguindo o [tutorial do pessoal do PSICODATA](https://medium.com/psicodata/baixando-e-processando-dados-do-datasus-sobre-suic%C3%ADdio-com-python-656afa17f6ad), você pode ver o notebook que utilizei para minerar os dados brutos [clicando aqui](https://colab.research.google.com/drive/1qo8HUYRPzXAugIZc5glOB8W_tX42QQ8k?usp=sharing) .


## Objetivo:

* Reduzir o tamanho da dataframe em ao menos 50%

## Metodologia
* Verificar o tamanho inicial do DataSet;
* Cortar colunas e dados menos importantes ou desnecessários;
* Trabalhar a tipagem dos dados;
* Exportar em CSV;
* Verificar o tamanho final do arquivo
* Calcular a porcentagem de redução dos dados em relação o dataset inicial.



In [1]:
import os
import pandas as pd

In [2]:
# Como baseline vamos verificando quais são as dimensões iniciais do arquivo
size = os.path.getsize("Dados/Nascimentos Brasil 2020 - DataSUS - Dados Brutos.csv")
tamanho_inicial = round(size/1e+6, 2)
print(f"Tamanho inicial do arquivo: {tamanho_inicial} MB")


Tamanho inicial do arquivo: 669.22 MB


## Análise do tamanho
669 Mega, é bastante peso para um arquivo, principalmente quando estamos falando em um .csv, vai dar trabalho para meu i5, sempre existe a possibilidade de utilizar o colab, que oferece uma capacidade de 12 gb de processamento no plano free, porém vou continuar local.

## Dimensões da tabela
Visto o tamanho do arquivo, agora vamos carregar o data-set e dar uma olhada nas dimensões da tabela:

In [9]:
df = pd.read_csv("Dados/Nascimentos Brasil 2020 - DataSUS - Dados Brutos.csv")
df.head()

C:\Users\manda\AppData\Local\Temp\ipykernel_11936\249228802.py:1: DtypeWarning: Columns (36) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("Dados/Nascimentos Brasil 2020 - DataSUS - Dados Brutos.csv")


Unnamed: 0 level_0  level_1  level_2  ORIGEM   CODESTAB  CODMUNNASC  \
0           0      AC     2020        0       1  6205224.0      110020   
1           1      AC     2020        1       1        NaN      110020   
2           2      AC     2020        2       1  3970442.0      110020   
3           3      AC     2020        3       1  5701929.0      120001   
4           4      AC     2020        4       1  5701929.0      120001   

   LOCNASC  IDADEMAE  ESTCIVMAE  ...  TPDOCRESP  DTDECLARAC  ESCMAEAGR1  \
0        1      34.0        2.0  ...        2.0   3042020.0         8.0   
1        4      18.0        1.0  ...        5.0   3052020.0         9.0   
2        1      26.0        2.0  ...        3.0   6022020.0         7.0   
3        1      29.0        5.0  ...        3.0   7012020.0         5.0   
4        1      20.0        5.0  ...        3.0  25012020.0         4.0   

   STDNEPIDEM  STDNNOVA  CODPAISRES  TPROBSON  PARIDADE  KOTELCHUCK  CONTADOR  
0         0.0         1         1.0         5         1           5      7565  
1         0.0         1         1.0        11         1           9      8023  
2         0.0         1         1.0         2         0           5      9017  
3         0.0         1         1.0         3         1           4     13104  
4         0.0         1         1.0         7         1           5     13105  

[5 rows x 65 columns]

In [4]:
print(f"Total de dados: {round(2730145*65/1000000,2)} mi")

Total de dados: 177.46 mi


### Demorou 52seg para carregar o dataframe, isso demonstra o imenso poder de processamento do `PYTHON`, já nao caberia nas 1.048.576 de linhas do `EXCEL`.  
### 177 milhões de entradas, será que podemos chamar de `BIG DATA`?

In [6]:
df.to_csv("Reduzindo SINASC 001.csv")

In [23]:
# Salvando lista com os cabeçalhos e index dos cabeçalhos
cabeçalhos = dict()
for index, coluna in enumerate(df.columns):
    cabeçalhos[index] = coluna
cabeçalhos


{0: 'Unnamed: 0',
 1: 'level_0',
 2: 'level_1',
 3: 'level_2',
 4: 'ORIGEM',
 5: 'CODESTAB',
 6: 'CODMUNNASC',
 7: 'LOCNASC',
 8: 'IDADEMAE',
 9: 'ESTCIVMAE',
 10: 'ESCMAE',
 11: 'CODOCUPMAE',
 12: 'QTDFILVIVO',
 13: 'QTDFILMORT',
 14: 'CODMUNRES',
 15: 'GESTACAO',
 16: 'GRAVIDEZ',
 17: 'PARTO',
 18: 'CONSULTAS',
 19: 'DTNASC',
 20: 'HORANASC',
 21: 'SEXO',
 22: 'APGAR1',
 23: 'APGAR5',
 24: 'RACACOR',
 25: 'PESO',
 26: 'IDANOMAL',
 27: 'DTCADASTRO',
 28: 'CODANOMAL',
 29: 'NUMEROLOTE',
 30: 'VERSAOSIST',
 31: 'DTRECEBIM',
 32: 'DIFDATA',
 33: 'DTRECORIGA',
 34: 'NATURALMAE',
 35: 'CODMUNNATU',
 36: 'CODUFNATU',
 37: 'ESCMAE2010',
 38: 'SERIESCMAE',
 39: 'DTNASCMAE',
 40: 'RACACORMAE',
 41: 'QTDGESTANT',
 42: 'QTDPARTNOR',
 43: 'QTDPARTCES',
 44: 'IDADEPAI',
 45: 'DTULTMENST',
 46: 'SEMAGESTAC',
 47: 'TPMETESTIM',
 48: 'CONSPRENAT',
 49: 'MESPRENAT',
 50: 'TPAPRESENT',
 51: 'STTRABPART',
 52: 'STCESPARTO',
 53: 'TPNASCASSI',
 54: 'TPFUNCRESP',
 55: 'TPDOCRESP',
 56: 'DTDECLARAC',
 57: 

In [38]:
# Limpando memória RAM
df = 0

# Reduzindo os dados
* Agora em posse do [dicionário dos cabeçalhos](https://diaad.s3.sa-east-1.amazonaws.com/sinasc/SINASC+-+Estrutura.pdf) e nossa lista atual, vamos refazer este dataframe visando eliminar algumas colunas, utilizando o parâmetro usecols.

Reduzi de 64 colunas para 43.

In [36]:
colunas = (1,6,7,8,9,10,11,12,13,14,15,16,18,19,20,
           21,22,23,24,25,26,28,35,36,37,38,39,40,41,
           42,43,44,46,48,49,51,52,53,57,60,61,62,63)

In [40]:
print(f"Total de dados: {round(2730145*43/1000000,2)} mi")

Total de dados: 117.4 mi


In [39]:
df = pd.read_csv("Dados/Nascimentos Brasil 2020 - DataSUS - Dados Brutos.csv", usecols=colunas)
df.head()

C:\Users\manda\AppData\Local\Temp\ipykernel_11936\3084648843.py:1: DtypeWarning: Columns (36) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("Dados/Nascimentos Brasil 2020 - DataSUS - Dados Brutos.csv", usecols=colunas)


level_0  CODMUNNASC  LOCNASC  IDADEMAE  ESTCIVMAE  ESCMAE  CODOCUPMAE  \
0      AC      110020        1      34.0        2.0     5.0    251605.0   
1      AC      110020        4      18.0        1.0     9.0    999992.0   
2      AC      110020        1      26.0        2.0     5.0    422105.0   
3      AC      120001        1      29.0        5.0     4.0    999992.0   
4      AC      120001        1      20.0        5.0     4.0    999992.0   

   QTDFILVIVO  QTDFILMORT  CODMUNRES  ...  CONSPRENAT  MESPRENAT  STTRABPART  \
0         1.0         1.0     120040  ...         9.0        2.0         2.0   
1         1.0         0.0     120040  ...         NaN        NaN         NaN   
2         NaN         NaN     120040  ...        10.0        1.0         1.0   
3         2.0         0.0     120001  ...         6.0        3.0         2.0   
4         0.0         1.0     120001  ...         7.0        2.0         2.0   

   STCESPARTO  TPNASCASSI  ESCMAEAGR1  CODPAISRES  TPROBSON  PARIDADE  \
0         1.0         1.0         8.0         1.0         5         1   
1         3.0         4.0         9.0         1.0        11         1   
2         2.0         1.0         7.0         1.0         2         0   
3         3.0         1.0         5.0         1.0         3         1   
4         3.0         2.0         4.0         1.0         7         1   

   KOTELCHUCK  
0           5  
1           9  
2           5  
3           4  
4           5  

[5 rows x 43 columns]

In [43]:
# Salvado csv com colunas reduzidas
df.to_csv('Dados/Nascimentos Brasil 2020 - colunas reduzidas - v 002.csv')